From 9fcdb1813c0a1fe049cec492675d521148f08a2f Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:45 -0400
Subject: [virt] kvm/x86: Processor mode may be determined incorrectly

Message-id: <1406040016-3289-191-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86269
O-Subject: [RHEL7 PATCH v2 190/221] KVM: x86: Processor mode may be determined incorrectly
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Nadav Amit <namit@cs.technion.ac.il>

If EFER.LMA is off, cs.l does not determine execution mode.
Currently, the emulation engine assumes differently.

Signed-off-by: Nadav Amit <namit@cs.technion.ac.il>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit 42bf549f3c672006ba18e97152cbc563315ba4e6)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 5b9a4f9..da571a9 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -4890,7 +4890,7 @@ static void init_emulate_ctxt(struct kvm_vcpu *vcpu)
  ctxt->eip = kvm_rip_read(vcpu);
  ctxt->mode = (!is_protmode(vcpu))  ? X86EMUL_MODE_REAL :
        (ctxt->eflags & X86_EFLAGS_VM) ? X86EMUL_MODE_VM86 :
-       cs_l    ? X86EMUL_MODE_PROT64 :
+       (cs_l && is_long_mode(vcpu)) ? X86EMUL_MODE_PROT64 :
        cs_db    ? X86EMUL_MODE_PROT32 :
          X86EMUL_MODE_PROT16;
  ctxt->guest_mode = is_guest_mode(vcpu);
-- 
1.7.1